### PR TITLE
MAINT: special: clean up factorial corner cases, including complex nans

### PR DESCRIPTION
**Pull request by h-vetinari**
This is part 2/N towards #18409, see #19989 for the predecessor and #19988 for an overview.

It contains minor behavioural changes around whether e.g. the dtype of an empty array is returned unchanged in all cases or whether we raise a type error on the dtype (more consistent with all other wrongly-typed inputs). It should be pretty minor stuff, and I'm happy to explain in more detail (beyond the code) what's happening, but for now I need to figure out how I can marry the complex-nan testing with the `xp_assert_*` infrastructure...

I reintroduced the `assert_really_equal` wrapper mainly for discussion purposes - it doesn't have to look like this, but IMO it'd be pretty benign, as all the core equality logic would continue to live in the `xp_assert_*` functions, and only the extra type- & nan-comparisons on top would live in that simple wrapper.

Or we could say we don't care about differences between float nans and complex nans (or whether a complex nan is `np.nan + 0j` or `np.nan + 1j*np.nan`). 🤷 

In short: the middle 3 commits are up for discussion and could be squashed/dropped/reshuffled. The first one is ready though - I can also break that out into a separate PR if desired.

CC @lucascolley @j-bowhay

### CODE REVIEW

**Comment by h-vetinari** [comment]
According to the release [plan](https://discuss.scientific-python.org/t/proposed-release-schedule-for-scipy-1-15-0/1502), we're branching in about 6 weeks, and I'd like to finish the factorial changes for 1.15. There's two PRs after this one still to come (draft preview of all changes in #19988).

I'm happy to hash out solutions for the situation here -- either have a thin wrapper around `xp_assert_*` or alternatively, ignore the specific representation of complex nans (e.g. `NaN + 0j` vs. `NaN + 1j * NaN`), or perhaps yet another solution I'm not seeing. Again, that I'm reinstating `assert_really_equal` is only for discussion purposes resp. getting this PR back up to speed.

Would appreciate some inputs @lucascolley @j-bowhay (or @steppi @mdhaber if interested)

**Comment by h-vetinari** [comment]
Gentle ping @steppi @lucascolley

**Comment by h-vetinari** [comment]
> According to the release [plan](https://discuss.scientific-python.org/t/proposed-release-schedule-for-scipy-1-15-0/1502), we're branching in about 6 weeks, and I'd like to finish the factorial changes for 1.15.

two out of those 6 weeks have now passed. Could I please get some inputs/preferences here @steppi @lucascolley?

Otherwise I'll move forward with the current state of this PR. The testing can always be relaxed later (if we decide to ignore different varieties of complex NaNs, for example).

**Review by lucascolley — APPROVED** [review]
LGTM!
